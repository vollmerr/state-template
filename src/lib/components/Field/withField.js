import withFieldWrapper from './withFieldWrapper';
import withReduxField from './withReduxForm';

// HOC for mapping a component to redux form format
// then passing to redux forms Field component
const withField = (emptyValidator) => (component, options) => {
  // attach aria attribute, error message, label, etc
  const WrappedField = withFieldWrapper(component, options);
  // map to redux form's Field
  return withReduxField(emptyValidator)(WrappedField);
};

export default withField;
