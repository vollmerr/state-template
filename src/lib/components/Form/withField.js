import React from 'react';
import { Field } from 'redux-form';

// HOC wrapper around redux-form's Field to alter props
// before they get to the compoonent
const withField = emptyValidator => component => (props) => {
  const { disabled, required, validate = [] } = props;
  // attach the validation for required based off input type
  if (required) {
    validate.push(emptyValidator);
  }
  // do not validate disabled fields
  const fixedValidate = !disabled ? validate : [];

  const fixedProps = {
    ...props,
    component,
    validate: fixedValidate,
  };

  return <Field {...fixedProps} />;
};

export default withField;
