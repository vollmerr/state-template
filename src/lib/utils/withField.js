import React from 'react';
import { Field } from 'redux-form';

// HOC wrapper around redux-form's Field to alter props
// before they get to the compoonent
const withField = emptyValidator => component => (props) => {
  const { disabled, required, validate = [] } = props;

  let toValidate;
  // only validate fields that are not disabled
  if (!disabled) {
    // attach the validation for required based off input type
    if (required) {
      toValidate = [...validate, emptyValidator];
    } else {
      toValidate = validate;
    }
  }

  const fieldProps = {
    ...props,
    component,
    validate: toValidate,
  };

  return <Field {...fieldProps} />;
};

export default withField;
