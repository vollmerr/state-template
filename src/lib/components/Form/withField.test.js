import React from 'react';
import { shallow } from 'enzyme';
import { Field } from 'redux-form';

import withField from './withField';

const { testUtils } = global;

const emptyValidator = jest.fn();
const component = 'input';
const withHOC = withField(emptyValidator);
const Input = withHOC(component);

const props = {
  input: testUtils.getInputProp(),
  meta: testUtils.getMetaProp(),
  label: 'test label',
  disabled: false,
  required: true,
  validate: [jest.fn],
};

let wrapper;
describe('withField', () => {
  beforeEach(() => {
    wrapper = shallow(<Input {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass the `component` to a redux-form `Field`', () => {
    const ReduxFormField = wrapper.find(Field);
    expect(ReduxFormField.length).toBe(1);
    expect(ReduxFormField.prop('component')).toEqual(component);
  });

  it('should add a empty validation if `required`', () => {
    const validation = wrapper.prop('validate').pop();
    expect(validation).toEqual(emptyValidator);
  });

  it('should ignore all validation if `disabled`', () => {
    wrapper.setProps({ disabled: true });
    expect(wrapper.prop('validate')).toBeFalsy();
  });

  it('should not add empty validation if not required', () => {
    wrapper.setProps({ required: false });
    const validations = wrapper.prop('validate');
    expect(validations.every(x => x !== emptyValidator)).toBeTruthy();
  });
});
