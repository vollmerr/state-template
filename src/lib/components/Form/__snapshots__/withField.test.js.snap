// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withField should render correctly 1`] = `
<Field
  component="input"
  disabled={false}
  input={
    Object {
      "name": "testName",
      "onChange": [MockFunction],
      "value": "test value",
    }
  }
  label="test label"
  meta={
    Object {
      "error": null,
      "touched": false,
    }
  }
  required={true}
  validate={
    Array [
      [Function],
      [MockFunction],
    ]
  }
/>
`;
